### ISSUE TITLE
MLX SFT ignores warmup, scheduler, and configured weight decay

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#686]
## Description

The MLX backend validates `warmup_ratio`, `scheduler`, and `weight_decay`, but constructs its
optimizer with only a scalar learning rate:

```python
optim.AdamW(learning_rate=float(cfg.training.lr))
```

No warmup/scheduler is built in this path, and configured weight decay is not passed.

Affected revision: `ef2e02b51a1b34546823edce55715d28234805cb` (Soup 0.74.0).

Relevant code:

- [optimizer construction](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/trainer/mlx_sft.py#L312-L330)
- [validated defaults and fields](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/config/schema.py#L1096-L1098)
- [scheduler field](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/config/schema.py#L1233)

## Steps to reproduce

Configure an MLX run with:

```yaml
training:
  lr: 0.0002
  weight_decay: 0.3
  warmup_ratio: 0.2
  scheduler: cosine
```

Intercept the real wrapper call to the optimizer constructor:

```text
forwarded kwargs: {"learning_rate": 0.0002}
```

## Expected behavior

The effective optimizer and learning-rate curve should match the validated Soup configuration,
or unsupported choices should be rejected clearly.

## Impact

MLX training silently runs a different recipe. Custom weight decay is lost, and a scalar LR
does not implement the configured warmup followed by cosine decay. This can affect stability,
convergence, and backend comparisons.

## Proposed fix and acceptance criteria

- [ ] Build the requested MLX schedule with a clearly defined optimizer-step unit.
- [ ] Pass configured weight decay.
- [ ] Map supported optimizer names and reject unsupported names instead of silently using AdamW.
- [ ] Test the observed LR at the first, warmup-boundary, and final updates.
- [ ] Store the effective optimizer/schedule configuration in adapter metadata.

## Environment

Argument-construction probe run under Python 3.12.2 against the original Soup wrapper. The
optimizer was replaced by a recorder; no numerical Metal update or final-quality claim is made.

**Comment by MakazhanAlpamys** [comment]
Re-confirmed still broken on current `main` after #696 and #698 land, via the real CLI. The optimizer really is constructed as `AdamW(learning_rate=0.0002)` and nothing else — a config setting `weight_decay: 0.3`, `warmup_ratio: 0.2` and `scheduler: cosine` has all three silently dropped.

One addition to the report: **none of the three appears in `_check_unsupported()`**, so that run printed no `MLX backend ignores:` line at all. They are validated by the schema, accepted, and then dropped without a word — which is this repo's second most common defect shape, a field that validates and is read by nothing.

That also gives the fix a cheap first half worth landing even if the full wiring is harder: adding the three to `_check_unsupported()` turns a silent drop into a warning, and it can be verified on any platform. The real wiring (an `mlx.optimizers` schedule for warmup/cosine, and `weight_decay=` on `AdamW`) needs Apple Silicon to verify behaviourally, though it can be pinned structurally with a fake the way #696 and #698 do it.

Unclaimed.

**Comment by Srinivasan8888** [comment]
Claiming this. I have the M1 8 GB from the #23 record, and I already have #683 open against this file, so I can keep the two from colliding.

## Reproduced before claiming, through the real wrapper

Not by reading the line — I replaced `mlx.optimizers.AdamW` with a subclass that records its kwargs and ran the actual `MLXSFTTrainerWrapper`:

```
config asked for : lr=0.0002  weight_decay=0.3  warmup_ratio=0.2  scheduler=cosine
forwarded kwargs : {'learning_rate': 0.0002}
```

All three dropped, and — as you say — no `MLX backend ignores:` line, so they are validated, accepted, and discarded silently. That is the shape.

## The measurement that decides the design

Your acceptance list asks for "a clearly defined optimizer-step unit", and that is the part worth settling before any code. I measured it rather than assuming:

```python
opt = optim.AdamW(learning_rate=sched)      # sched records the step it is called with
for i in range(4): opt.update(m, grads)
-> schedule called with steps: [0, 0, 1, 2, 3]
-> opt.step after 4 updates  : 4
```

`opt.step` advances once per `optimizer.update()`, and mlx-lm calls that only when `it % grad_accumulation_steps == 0`. **So the schedule's unit is optimizer updates, not iterations** — and the total it must be built against is `iters // gradient_accumulation_steps`, not `iters`. Building it against `iters` would stretch the warmup by a factor of `accum` and leave the cosine never reaching its floor, which is exactly the kind of silently-plausible curve this issue is about. I will pin that arithmetic with a test rather than a comment.

MLX has the pieces, so this needs no hand-rolled math:

```
join_schedules([linear_schedule(0, lr, warmup), cosine_decay(lr, total - warmup)], [warmup])
-> at steps 0, 5, 10, 25, 50 with lr=2e-4, warmup=10, total=50:
   [0.0, 0.0001, 0.0002, 0.00013827, 0.0]
```

Peak exactly at the warmup boundary, zero at the end. `AdamW`'s `learning_rate` already accepts `Union[float, Callable]`, and `weight_decay` is a plain kwarg it has always had.

## What I intend, against your criteria

- **schedule** built in optimizer-update units, from `scheduler` + `warmup_ratio`; observed LR tested at the first update, the warmup boundary and the final update, as you asked
- **weight decay** passed — currently the schema default 0.01 is what every MLX run silently gets, whatever the config says
- **optimizer + scheduler names mapped explicitly, unsupported ones rejected, not silently AdamW'd.** MLX ships `AdamW / Adam / SGD / Lion / Adafactor / Adagrad / Adamax / AdaDelta / RMSprop / Muon`, and the schema's default is `adamw_torch` against an allowlist (`optimizer_zoo`) far wider than that. So most valid Soup optimizer names have no MLX equivalent, and today all of them silently become AdamW. I will map the ones that genuinely correspond and refuse the rest by name.
- **`_check_unsupported()`** gains only what remains genuinely unsupported after this — a warning for something I just wired would be worse than none
- **adapter metadata** records the effective optimizer and schedule

One thing I will decide by measurement rather than by preference, and report either way: whether `warmup_ratio` rounding to zero updates on a short run (`0.03 * 12 = 0`) should be a silent no-warmup or a warning. My lean is a warning, because a benchmark-sized run silently getting no warmup is how the #716 class of bug happens.

## Overlap, named rather than out-run

`src/soup_cli/trainer/mlx_sft.py` currently has three of us near it:

- **@AmirF194** — #698 (`gradient_checkpointing`), open, touches the `TrainingArgs(...)` call.
- **me** — #733 (#683, response-only masking), open, touches the dataset construction and the `train()` call.
- **this** — the optimizer construction, which is a third region and adjacent to both.

**I will rebase onto #698 when it lands, and I will not touch the `TrainingArgs(...)` call.** @AmirF194 — if you would rather take #686 as the natural continuation of #684/#685, say so and I will hand it over untouched; you have been working this family longer than I have. I will hold off pushing anything for a while to give you the chance to say so.

**Comment by MakazhanAlpamys** [comment]
Closed by #734, merged as `765f5b4c`. Credit to @Srinivasan8888, who has the M1 this needed and reproduced the defect through the real wrapper before claiming.

**What shipped.** `training.warmup_ratio`, `training.scheduler`, `training.weight_decay` and `training.optimizer` are now planned by `plan_optimizer` and built by `build_optimizer` on the MLX path, with the resolved values reaching `adapter_config.json` through `as_metadata()`. Before this, `build_optimizer` constructed `optim.AdamW(learning_rate=float(cfg.training.lr))` and dropped all four, so a config asking for cosine decay with warmup trained at a flat rate and said nothing.

**Nine mutations at merge, nine killed**, every one on an interpreter where `find_spec("mlx") is None` — which is the condition every CI cell runs under, and the condition the earlier rounds' survivors were hiding behind. That includes `build_optimizer` restored to this issue's defect verbatim (12 tests) and the `constant` / `constant_with_warmup` branch disabled so those schedulers silently decay to zero while the metadata still records `constant` (2 tests, both spellings).

**The reusable lesson from the review**, recorded here because it is not specific to MLX: the test that would have caught the last survivor already existed and asserted exactly the right property — but it evaluated the schedule through a helper opening with `pytest.importorskip("mlx.core")`, so it was skipped on every runner this project has. The builder-level control meant to cover for it passed `warmup_ratio=0.0`, which returns before the MLX import, so its assertion held for every possible implementation of the branch it was named after. A hardware-gated test plus a control that cannot discriminate is the same as no coverage, and it looks green.

`benchmarks/harness/mlx_sft_smoke.py` pins `scheduler`, `warmup_ratio` and `weight_decay`, so the published record is not silently invalidated by this change — the contributor pre-empted that rather than being asked.
